Reorder Pauli terms before Trotterization

altaris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Reordering the terms of a Pauli operator can reduce the depth of the evolution circuit if using the Lie-Trotter or Suzuki-Trotter methods. This PR adds the option to do just that.

### Details and comments

This PR adds 
* the method `synthesis.evolution.product_formula.reorder_paulis` that reorders Pauli terms based on a greedy graph coloring heuristic;
* the `reorder` option in the `ProductFormula`, `LieTrotter` and `SuzukiTrotter` constructors to allow for reordering before synthesis.

altaris committed:
Added option to reorder Paulis before Trotterization

Reordering the terms of a Pauli operator can reduce the depth of the evolution
circuit if using the Lie-Trotter or Suzuki-Trotter methods. This commit adds a
method `synthesis.evolution.product_formula.reorder_paulis` that does just that,
based on a greedy graph coloring heuristic. This commit also adds the `reorder` option in
the `ProductFormula`, `LieTrotter` and `SuzukiTrotter` constructors to allow for
reordering before synthesis.

altaris committed:
Added unit tests for Pauli reordering

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=12925) <br/>All committers have signed the CLA.

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,64 @@ def evolve_pauli(
+                break
+    # TODO: The graph is likely not very big, so an exact coloring could be
+    # computed in a reasonable time.
+    coloring = rx.graph_greedy_color(graph)
```
A quick comment: the Rustworkx's `graph_greedy_color` function can take an optional argument `strategy` that specifies the heuristic coloring method used, please see https://www.rustworkx.org/dev/apiref/rustworkx.ColoringStrategy.html for detail. You might want to experiment with the different strategies available to see which performs best for this application (or run all three one after another and choose the result with the smallest number of colors).

Cryoris replied:
I agree, we should try this or expose an argument allowing to set the coloring strategy.

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,64 @@ def evolve_pauli(
+                graph.add_edge(idx1, idx2, None)
+                break
+    # TODO: The graph is likely not very big, so an exact coloring could be
+    # computed in a reasonable time.
```
Out of curiosity, how many nodes are typically there and how much is it important to get the optimum number of colors? (For instance, there are standard ways to translate a graph coloring problem into SAT which could then be solved with say Z3).

Cryoris commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -78,6 +78,7 @@ def __init__(
             | None
         ) = None,
         wrap: bool = False,
+        reorder: bool = False,
```
We can discuss this, but I think this should be `True` per default as it can significantly optimize the circuit depth 🙂 It will change the output of the current circuits, but as this is a synthesis routine which only promises to output a circuit implementing the Trotter formula, changes like this are fine I think.

sbrandhsn replied:
I disagree. We should only consider performing Pauli term reordering by default for a given Hamiltonian, if we
- understand the potential classical runtime implications
- (strongly) suspect we can gain reductions in circuit depth by inspecting the Hamiltonian
- care about circuit depth at all as specified by some transpilation option
- leveraged Rust for getting the most efficient implementation
The underlying combinatorial problem is NP-complete and while heuristics yield a great runtime improvement they can still be quite slow compared to the rest of the transpilation steps, especially on larger Hamiltonians. Otherweise, I love the PR, thanks @altaris !

I'm mentioning @ajavadia in case he wants to chime in. :-)

Cryoris replied:
Runtime cost is a good point 👍🏻 it would be interesting to see the that for large scale Hamiltonians, as in most cases they have an underlying structure. Maybe we could check some from the Hamlib? 🙂

ShellyGarion replied:
Is there a benchmark such as Benchpress HamLib https://github.com/Qiskit/benchpress/tree/main that indicates the CX cost/depth improvement and the time consumption ?

Cryoris replied:
Conclusion: we're not enabling this per default for now.

Cryoris commented:
@altaris are you still pursuing this or do would you like us to help? 🙂

altaris commented:
Sure! I'd love to see it through 💪

altaris committed:
reorder_paulis acts as the identity if the operator has only one term

Before that, if reorder_paulis was given a single-term SparsePauliOp object, it
would return it encapsulated in a list. Now, it returns it as is.

altaris committed:
Reorder Pauli terms by default

altaris committed:
reorder_paulis is deterministic and order-invariant

altaris committed:
Fixed SuzukiTrotter unit tests

altaris committed:
No Pauli reordering by default

altaris committed:
Added Pauli reordering option in QDrift

Cryoris commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
+    This method works in three steps. First, a graph is constructed, where the
+    nodes are the terms of the operator and where two nodes are connected if
+    their term acts on the same qubit (for example, the terms $IXX$ and $IYI$
+    would be connected, but not $IXX$ and $YII$). Then, the graph is colored.
```
This is not API documented, so this is just as info: to render math in our docs we have to use 
```
:math:`XII`
```
instead of `$XII$` 🙂

Cryoris commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
+        operators = [operators]
+
+    terms = sum((o.to_list() for o in operators), [])
+    terms = sorted(terms, key=_term_sort_key)
```
Why do we need to sort the operators? To enable a deterministic outcome?

Cryoris commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
+        # operators has a single SparsePauliOp, which has > 1 terms
+        or (len(operators) == 1 and len(operators[0]) > 1)
+    ):
+        return operators
```
Pauli operators that are given in a list all commute, e.g. we would give
```
[SparsePauliOp(["ZZI", "IZZ", "ZII"]), SparsePauliOp(["YII", "YYY"])]
```
This information should be preserved if provided by the users 🙂 So I think we should simply do a re-ordering per element in the list, instead of merging all operators and discarding this initial information. If users would like to merge all operators, this could simply be done by adding the operators before passing them into the evolution.

mrossinek replied:
I agree that preserving this information is needed. I suppose you could simply call the algorithm below on each entry of this list separately and then return a new list of operators.

Cryoris commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
+    for (idx1, term1), (idx2, term2) in combinations(indexed_nodes, 2):
+        # Add an edge between term1 and term2 if they touch the same qubit
+        for a, b in zip(term1[0], term2[0]):
+            if not (a == "I" or b == "I"):
```
```suggestion
            if a != "I" and b != "I":
```
should be slightly faster since the expression can terminate already if the first part of the `and` statement is `False` 🙂

Cryoris reviewed (commented):
Overall the PR looks very nice! I left some comments below, in particular about the handling of lists of operators. Could you also add a release note describing the new argument? 🙂

mrossinek commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -177,6 +186,79 @@ def evolve_pauli(
+        for a, b in zip(term1[0], term2[0]):
+            if not (a == "I" or b == "I"):
+                graph.add_edge(idx1, idx2, None)
+                break
```
I would think there is a better way to do this based on the `PauliList` boolean array saved inside the `SparsePauliOp` instances. A simple way is the following:
```python
x_or_z = op.paulis.x | op.paulis.z
overlaps = x_or_z @ x_or_z.T
```
This yields a matrix indicating on entry `(i, j)` whether the Pauli term at index `i` has overlap with that at index `j`.

This computes the full matrix but you are really only interested in the upper triangular indices. So there may be an even more efficient way to do this...

mrossinek replied:
Actually, this might also help with the algorithm becoming deterministic without requiring the terms to be sorted first :thinking:

Cryoris replied:
This is now obsolete as we deal with sparse pauli labels 🙂

mrossinek reviewed (commented):
Just adding my 2 cents here

altaris committed:
Typset math correctly

altaris committed:
Reorder ops individually

If `reorder_paulis` is given a list of operators, then they are reordered and returned individually. Before this co
mmit, they used to be added all together and then reordered, which always resulted in a single operator.

altaris committed:
Option to change graph coloring heuristic

alexanderivrii commented:
A quick question. The overall consensus is that "reorder" should be "False" by default? Or that it should be "True" if reordering is sufficiently fast?

Cryoris committed:
Merge branch 'main' into reorder-paulis

Cryoris committed:
add plugin

Cryoris committed:
consistent with rustiq

Cryoris committed:
rm print, fix sorting

Cryoris committed:
lint & docs

ShellyGarion reviewed (commented):
LGTM. thanks for the nice contribution to Qiskit. Some release notes are needed.

Cryoris commented on qiskit/synthesis/evolution/qdrift.py:
```diff
@@ -125,4 +132,8 @@ def expand(self, evolution: PauliEvolutionGate) -> list[tuple[str, tuple[int], f
             (pauli[0], pauli[1], np.real(np.sign(pauli[2])) * rescaled_time) for pauli in sampled
         ]
+
+        if self.preserve_order:
```
Only re-order if the order is _not_ preserved 🙂 
```suggestion
        if not self.preserve_order:
```

Cryoris committed:
Fix QDrift reordering

alexanderivrii commented on qiskit/circuit/library/pauli_evolution.py:
```diff
@@ -143,7 +149,14 @@ def time(self, time: ParameterValueType) -> None:
+
+            synthesis = LieTrotter()
+        else:
+            synthesis = self.synthesis
```
Do we need this, i.e. shouldn't `self.synthesis` always return a non-`None` synthesis method?

Cryoris replied:
Correct, fixed in 56e619c

alexanderivrii commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -79,8 +80,19 @@ def __init__(
                 built.
             wrap: Whether to wrap the atomic evolutions into custom gate objects. This only takes
                 effect when ``atomic_evolution is None``.
+            preserve_order: Whether to allow preserve_ordering the terms of the operator to
```
"preserve_order" allows "preserve_ordering" 😄 you might want to rephrase this slightly

Cryoris replied:
Hehe, this happened due to `%s/reorder/preserve_order` 😅

Cryoris replied:
Fixed in 56e619c

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -84,11 +90,15 @@ def __init__(
                 this to ``True`` is slower than ``False``. This only takes effect when
                 ``atomic_evolution is None``.
+            preserve_order: Whether to allow preserve_ordering the terms of the operator to
+                potentially yield a shallower evolution circuit. Not relevant
```
and here too

Cryoris replied:
Fixed in 56e619c

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -239,3 +250,74 @@ def real_or_fail(value, tol=100):
+
+
+def reorder_paulis(
+    paulis: SparsePauliLabel, strategy: rx.ColoringStrategy = rx.ColoringStrategy.Saturation
```
Out of curiosity, why did you choose this specific coloring strategy? 

One other thing you can do is to first call `rx.graph_bipartite_edge_color`, and if the graph happens to be bipartite (which is the case for many structured graphs of interest, like heavy-hexes), then we will automatically get a coloring with the minimum number of colors. And if the graph is not bipartite, we could use the greedy coloring function.

UPDATE: sorry, my comment above makes no sense whatsoever, in this algorithm you need a _node coloring_ algorithm, while the function I mentioned is for _edge coloring_.

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -239,3 +250,74 @@ def real_or_fail(value, tol=100):
+    paulis: SparsePauliLabel, strategy: rx.ColoringStrategy = rx.ColoringStrategy.Saturation
+) -> SparsePauliOp | list[SparsePauliOp]:
+    r"""
+    Creates an equivalent operator by preserve_ordering terms in order to yield a
```
At this point I am starting to doubt myself, and think that `preserve_ordering` is a valid English word.

Cryoris replied:
without underscore I think it is, but it's wrong here 😂

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -239,3 +250,74 @@ def real_or_fail(value, tol=100):
+
+    This method works in three steps. First, a graph is constructed, where the
+    nodes are the terms of the operator and where two nodes are connected if
+    their term acts on the same qubit (for example, the terms :math:`IXX` and
```
"their term" -> "their terms"

Cryoris replied:
Fixed in 56e619c

alexanderivrii commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -239,3 +250,74 @@ def real_or_fail(value, tol=100):
+        term = graph.nodes()[term_idx]
+        terms_by_color[color].append(term)
+
+    terms = sum(terms_by_color.values(), [])
```
This is probably obvious, but how is `sum` used here exactly? And why at the end `terms` is of the form  `SparsePauliOp | list[SparsePauliOp]`?

Cryoris replied:
It misuses `sum` to concatenate lists -- it's better to use `itertools.chain` though, I'll update

Cryoris replied:
Fixed in 56e619c

alexanderivrii commented on qiskit/synthesis/evolution/suzuki_trotter.py:
```diff
@@ -104,6 +105,9 @@ def __init__(
                 effect when ``atomic_evolution is None``.
+            preserve_order: Whether to allow reordering the terms of the operator to
+                potentially yield a shallower evolution circuit. Not relevant
+                when synthesizing operator with a single term. Defaults to ``True``.
```
Now I like the docstring 😄 

Actually, the way this is worded, it feels the other way around, i.e. `preserve_order` restricts the order to the original order.

Cryoris replied:
Fixed in 56e619c

alexanderivrii reviewed (commented):
Thanks @Cryoris  and the original contributor @altaris! I think allowing Pauli reordering is a useful option for the `PauliEvolutionGate` synthesis algorithms.

I have left a few minor comments and questions inline.

One more question: do you think that for Rustiq-based synthesis algorithm (and the plugin already accepts "preserve_order"), it might be beneficial to rearrange Paulis before passing the pauli network to Rustiq?

Cryoris commented:
> One more question: do you think that for Rustiq-based synthesis algorithm (and the plugin already accepts "preserve_order"), it might be beneficial to rearrange Paulis before passing the pauli network to Rustiq?

That is already possible to do with this PR, since the re-ordering happens before `expand`. So we can expose this in the Rustiq plugin, just like in the default plugin. So we would re-order with the graph coloring and then allow Rustiq to also re-order? Or are they redundant together?

alexanderivrii commented:
> That is already possible to do with this PR, since the re-ordering happens before expand. So we can expose this in the Rustiq plugin, just like in the default plugin. So we would re-order with the graph coloring and then allow Rustiq to also re-order? Or are they redundant together?

I think it makes sense to expose this to Rustiq plugin as well. And I think it makes sense to let Rustiq do the additional reordering.

Cryoris committed:
review comments

- add reno
- docs
- safer list concat

Cryoris committed:
Merge branch 'reorder-paulis' of github.com:altaris/qiskit into reorder-paulis

Cryoris committed:
Merge branch 'main' into reorder-paulis

Cryoris commented:
56e619c includes this option in the Rustiq plugin too 👍🏻

Cryoris committed:
rm dead code

mtreinish commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -60,6 +60,7 @@ def __init__(
             | None
         ) = None,
         wrap: bool = False,
+        preserve_order: bool = False,
```
I thought the consensus here was to not reorder terms by default because of the potential performance overhead. Shouldn't the default be to `True`?

mtreinish commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -63,6 +69,7 @@ def __init__(
             | None
         ) = None,
         wrap: bool = False,
+        preserve_order: bool = False,
```
Same question here

mtreinish replied:
```suggestion
        preserve_order: bool = True,
```

mtreinish commented on qiskit/synthesis/evolution/qdrift.py:
```diff
@@ -68,6 +68,7 @@ def __init__(
         ) = None,
         seed: int | None = None,
         wrap: bool = False,
+        preserve_order: bool = True,
```
This one defaults to `True` that's just inconsistent, but I think the expected value.

alexanderivrii commented on qiskit/transpiler/passes/synthesis/hls_plugins.py:
```diff
@@ -1528,6 +1535,9 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
 
+        if "preserve_order" in options and isinstance(algo, ProductFormula):
+            algo.preserve_order = options["preserve_order"]
+
```
Thanks for adding this. A very small nitpick: we don't need another call to `isinstance(also, ProductFormula)` here, since we've just checked this on the line above.

mtreinish commented:
I think given the open questions and the proximity to 1.3.0 (ie this is the last PR being tracked) it's probably better to defer this to 2.0.0 rather than force it in. This feature is a nice to have, but I don't think it's absolutely necessary for 1.3.0

raynelfss commented:
Pushing to 2.0.0 after further discussion...

Cryoris committed:
fix default value

Cryoris committed:
missed one

thanks Matthew!

mtreinish committed:
Merge branch 'main' into reorder-paulis

alexanderivrii reviewed (approved):
Thanks @Cryoris. This PR very nicely concludes the work that was done to improve the support for Pauli evolutions: representing them as gates, improving expansion methods, adding the plugin interface, new synthesis algorithms and new options, porting relevant code to Rust.

mtreinish commented on qiskit/synthesis/evolution/lie_trotter.py:
```diff
@@ -79,8 +80,19 @@ def __init__(
                 built.
             wrap: Whether to wrap the atomic evolutions into custom gate objects. This only takes
                 effect when ``atomic_evolution is None``.
+            preserve_order: If ``False``, allows reordering the terms of the operator to
+                potentially yield a shallower evolution circuit. Not relevant
+                when synthesizing operator with a single term.
```
I think from a docs perspective it would be good to explain the tradeoffs here, especially around runtime performance.

mtreinish commented on qiskit/synthesis/evolution/product_formula.py:
```diff
@@ -239,3 +251,61 @@ def real_or_fail(value, tol=100):
+    terms = sorted(paulis, key=_term_sort_key)
+    graph = rx.PyGraph()
+    graph.add_nodes_from(terms)
+    indexed_nodes = list(enumerate(graph.nodes()))
```
I think you don't actually need to cast this is as a list, it should all work fine without it and save a copy.

mtreinish reviewed (commented):
I'm fine with merging this as is. I just have 2 small nits inline, but they can be fixed in a followup without issue and aren't worth blocking this over.
